miner/algo_greedy_buckets.go

```diff
@@ -0,0 +1,228 @@
+				if receipt != nil {
+					// if the receipt is nil we don't attempt to retry the transaction - this is to mitigate abuse since
+					// without a receipt the default profit calculation for a transaction uses the gas limit which
+					// can cause the transaction to always be first in any profit-sorted transaction list
+					gasUsedMap[order] = receipt.GasUsed
+					CheckRetryOrderAndReinsert(order, orders, retryMap, retryLimit)
+				}
+				continue
+			}
+
+			if skip == shiftTx {
+				orders.ShiftAndPushByAccountForTx(tx)
+			}
+
+			effGapPrice, err := tx.EffectiveGasTip(envDiff.baseEnvironment.header.BaseFee)
+			if err == nil {
+				log.Trace("Included tx", "EGP", effGapPrice.String(), "gasUsed", receipt.GasUsed)
+			}
+		} else if bundle := order.Bundle(); bundle != nil {
+			err := envDiff.commitBundle(bundle, b.chainData, b.interrupt, algoConf)
+			if err != nil {
+				log.Trace("Could not apply bundle", "bundle", bundle.OriginalBundle.Hash, "err", err)
+				CheckRetryOrderAndReinsert(order, orders, retryMap, retryLimit)
+				continue
+			}
+
+			log.Trace("Included bundle", "bundleEGP", bundle.MevGasPrice.String(),
+				"gasUsed", bundle.TotalGasUsed, "ethToCoinbase", ethIntToFloat(bundle.TotalEth))
+			usedBundles = append(usedBundles, *bundle)
+		} else if sbundle := order.SBundle(); sbundle != nil {
+			usedEntry := types.UsedSBundle{
+				Bundle: sbundle.Bundle,
+			}
+			err := envDiff.commitSBundle(sbundle, b.chainData, b.interrupt, b.builderKey, algoConf)
+			if err != nil {
+				log.Trace("Could not apply sbundle", "bundle", sbundle.Bundle.Hash(), "err", err)
+				if ok := CheckRetryOrderAndReinsert(order, orders, retryMap, retryLimit); !ok {
+					usedEntry.Success = false
+					usedSbundles = append(usedSbundles, usedEntry)
+				}
+				continue
+			}
+
+			log.Trace("Included sbundle", "bundleEGP", sbundle.MevGasPrice.String(), "ethToCoinbase", ethIntToFloat(sbundle.Profit))
+			usedEntry.Success = true
+			usedSbundles = append(usedSbundles, usedEntry)
+		} else {
+			// note: this should never happen because we should not be inserting invalid transaction types into
+			// the orders heap
+			panic("unsupported order type found")
+		}
+	}
+	return usedBundles, usedSbundles
+}
+
+func (b *greedyBucketsBuilder) mergeOrdersIntoEnvDiff(
+	envDiff *environmentDiff, orders *types.TransactionsByPriceAndNonce) ([]types.SimulatedBundle, []types.UsedSBundle,
+) {
+	if orders.Peek() == nil {
+		return nil, nil
+	}
+
+	const retryLimit = 1
+
+	var (
+		baseFee      = envDiff.baseEnvironment.header.BaseFee
+		retryMap     = make(map[*types.TxWithMinerFee]int)
+		usedBundles  []types.SimulatedBundle
+		usedSbundles []types.UsedSBundle
+		transactions []*types.TxWithMinerFee
+		percent      = new(big.Float).Quo(
+			new(big.Float).SetInt(b.algoConf.ProfitThresholdPercent),
+			new(big.Float).SetInt(common.Big100),
+		)
+
+		CutoffPriceFromOrder = func(order *types.TxWithMinerFee, cutoffPercent *big.Float) *big.Int {
```

Comment:
Do you maybe want to move the closures into regular standalone functions for better readability?

Reply:
moved some of the closures into standalone functions